### PR TITLE
[SYCL][NFC] Avoid nullptr dereferencing of YDimExpr and ZDimExpr

### PR DESCRIPTION
**Pull request by schittir**
Add assert that arguments of work_group_size attributes cannot be nullptr

Klocwork exposed a (false positive) bug that nullptr derferencing of 
YDimExpr and ZDimExpr is possible for non sycl:: usage.
This is not practically possible because such (OpenCL and cl::) spellings
of the attribute require three arguments.

This patch
1.  Removes the ability of SetDefaultValue lambda to return nullptr
2. Asserts in SetDefaultValue that it is not possible for sycl:: case 
to have NULL arguments
3.  Moves error checking for three arguments for OpenCL and cl:: cases 
ahead of the assert 
4. Removes check for intel::reqd_work_group_size spelling

### CODE REVIEW

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 1]
```diff
@@ -3177,25 +3177,32 @@ static void handleWorkGroupSize(Sema &S, Decl *D, const ParsedAttr &AL) {
 
   ASTContext &Ctx = S.getASTContext();
 
-  if (!XDimExpr->isValueDependent() && !YDimExpr->isValueDependent() &&
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaDeclAttr.cpp:3180]: This is doing something subtly different than before. Before, if any of the expressions were value dependent, we didn't bother trying to verify any of the constant expressions. Now, we consider each dimension independently, which will change testing behavior (we can get diagnostics when we previously did not).

I'm not saying this is a bad change in behavior, but it is a difference in behavior and it's possible the CI pipeline is failing because of it.

Btw, this suggests we're missing test coverage that should be added as part of the patch. Can you add a test case that would crash without your changes so that we can verify that your fix doesn't regress some day in the future?

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 2]
```diff
@@ -3177,25 +3177,32 @@ static void handleWorkGroupSize(Sema &S, Decl *D, const ParsedAttr &AL) {
...
+  }
 
+  if (ZDimExprIsDereferencable) {
+    ExprResult ZDim = S.VerifyIntegerConstantExpression(ZDimExpr, &ZDimVal);
     if (ZDim.isInvalid())
       return;
     ZDimExpr = ZDim.get();
+  }
 
+  if ((XDimVal >= 1) && (YDimVal >= 1) && (ZDimVal >= 1)) {
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaDeclAttr.cpp:3205]: Why are you checking for `>= 1` now?

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 3]
```diff
@@ -3177,25 +3177,32 @@ static void handleWorkGroupSize(Sema &S, Decl *D, const ParsedAttr &AL) {
...
   ASTContext &Ctx = S.getASTContext();
 
-  if (!XDimExpr->isValueDependent() && !YDimExpr->isValueDependent() &&
-      !ZDimExpr->isValueDependent()) {
-    llvm::APSInt XDimVal, YDimVal, ZDimVal;
-    ExprResult XDim = S.VerifyIntegerConstantExpression(XDimExpr, &XDimVal);
-    ExprResult YDim = S.VerifyIntegerConstantExpression(YDimExpr, &YDimVal);
-    ExprResult ZDim = S.VerifyIntegerConstantExpression(ZDimExpr, &ZDimVal);
+  bool YDimExprIsDereferencable = YDimExpr && (!YDimExpr->isValueDependent());
+  bool ZDimExprIsDereferencable = ZDimExpr && (!ZDimExpr->isValueDependent());
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaDeclAttr.cpp:3181]:
```suggestion
  bool YDimExprIsDereferencable = YDimExpr && !YDimExpr->isValueDependent();
  bool ZDimExprIsDereferencable = ZDimExpr && !ZDimExpr->isValueDependent();
```
Parens don't change the order of evaluation here, so they can be removed.

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 2 above)
Comment (schittir) [reply clang/lib/Sema/SemaDeclAttr.cpp:3205]: I did that instead of checking that `(XDimExprIsDereferencable && YDimExprIsDereferencable && ZDimExprIsDereferencable) == 1` Because `XDimVal`, `YDimVal` and `ZDimVal` are definitely >= 1 in cases where `YDimExpr` and `ZDimExpr` are not `nullptr`

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 3 above)
Comment (schittir) [reply clang/lib/Sema/SemaDeclAttr.cpp:3181]: Ok.

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 1 above)
Comment (schittir) [reply clang/lib/Sema/SemaDeclAttr.cpp:3180]: I'm trying to figure out exactly how this behaviour is different..

Will add a test such that YDimExpr and ZDimExpr will be nullptr in the case where sycl:: or intel:: reqd_work_group_size spelling is not used. Sounds ok?

Comment (AaronBallman) [reply clang/lib/Sema/SemaDeclAttr.cpp:3180]: > I'm trying to figure out exactly how this behaviour is different..

It differs because the previous code would not verify the integer constant expressions of `y` or `z` unless all of x, y, and z are not value dependent. In the new code, we check them individually, so if x is not value dependent, but y and z are, we'll get errors before template instantiation instead of after (which may or may not be a bad thing!).

> Will add a test such that YDimExpr and ZDimExpr will be nullptr in the case where sycl:: or intel::reqd_work_group_size spelling is not used. Sounds ok?

That sounds sensible to me.

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 2 above)
Comment (AaronBallman) [reply clang/lib/Sema/SemaDeclAttr.cpp:3205]: Okay, I think I see what you're doing now, thanks! The stuff you are guarding can only be called if we have the x, y, and z dims available (they're not value dependent), so you're sort of trying to check the same thing, but in a different way than before. I think the old approach was a bit more clear -- it's not that we care about the value itself, it's that we care there *is* a value.

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 1 above)
Comment (erichkeane) [reply clang/lib/Sema/SemaDeclAttr.cpp:3180]: Right, the idea is you want to only do the checking if the three dimensions are either null or not-value-dependent, otherwise you should skip the checking.  Then, make sure you check them each individually, but only if they are non-null.

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 2 above)
Comment (erichkeane) [reply clang/lib/Sema/SemaDeclAttr.cpp:3205]: I don't think you can count on the value here, the user could insert a '1' here as well.

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 1 above)
Comment (schittir) [reply clang/lib/Sema/SemaDeclAttr.cpp:3180]: @erichkeane Are you suggesting a reversal of my order of operations, i.e., doing the individual checking within the if(!XDimExpr->isValueDependent() && YDimExprIsDereferencable && ZDimExprIsDerferencable) ?

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 2 above)
Comment (schittir) [reply clang/lib/Sema/SemaDeclAttr.cpp:3205]: Ok. Will change it back to the old approach. Thanks!

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 1 above)
Comment (erichkeane) [reply clang/lib/Sema/SemaDeclAttr.cpp:3180]: there is a top level check before all the checking to make sure that everything is either null or not-dependent. So something like:

    if ((XDimExpr && XDimExpr->isValueDependent()) | |YDimExpr && YDimExpr->isValueDependent()) || ZDimExpr && ZDimExpr->isValueDependent()))
        //skip checking.


Then when inspecting each value, you only check it if non-null.

Comment (schittir) [reply clang/lib/Sema/SemaDeclAttr.cpp:3180]: @erichkeane 
Couldn't find a case where YDimExpr or ZDimExpr can actually be nullptr: 

1. For non sycl:: or non intel:: use, such as the OpenCL spelling __attribute__((reqd_work_group_size(X,Y,Z)), three arguments are mandatory. 
2. For sycl:: or intel:: usage, when the user doesn't pass Y or Z arguments, there is a default initialization to 1
3. The possibility of nullptr in SetDefaultValue definition seems only theoretical

Comment (AaronBallman) [reply clang/lib/Sema/SemaDeclAttr.cpp:3180]: Agreed; I think the Klokwork report is a false positive, but we could eliminate it (hopefully) and improve the readability of the code by using an assert within the lambda to assert that the only way to call the lambda is for an attribute that *can* elide the Y and Z arguments. Then we can remove the return of `nullptr` from the lambda and simplify that code a bit.

Comment (erichkeane) [reply clang/lib/Sema/SemaDeclAttr.cpp:3180]: In that case, then yes, I agree with Aaron: We need to make it so that lambda doesn't have the ability to return nullptr.

Comment (schittir) [reply clang/lib/Sema/SemaDeclAttr.cpp:3180]: Modified the lambda such that it will not return a nullptr; added an assert.

**Review by erichkeane — APPROVED** [review]

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 4]
```diff
@@ -3148,13 +3148,13 @@ static void handleWorkGroupSize(Sema &S, Decl *D, const ParsedAttr &AL) {
...
-          AL.getScopeName()->isStr("intel")))
-            ? IntegerLiteral::Create(S.Context, llvm::APInt(32, 1),
-                                     S.Context.IntTy, AL.getLoc())
-            : nullptr;
+    assert((AL.getKind() == ParsedAttr::AT_ReqdWorkGroupSize && AL.hasScope() &&
+            (AL.getScopeName()->isStr("sycl") ||
+             AL.getScopeName()->isStr("intel"))) &&
+           "Attribute does not exist in sycl:: or intel:: scope");
+
+    Expr *E = IntegerLiteral::Create(S.Context, llvm::APInt(32, 1),
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaDeclAttr.cpp:3156]: Rather than create the literal, storing it into `E`, and then returning `E`, just return the literal directly (no local variable).

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 5]
```diff
@@ -3176,7 +3176,6 @@ static void handleWorkGroupSize(Sema &S, Decl *D, const ParsedAttr &AL) {
   }
 
   ASTContext &Ctx = S.getASTContext();
-
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaDeclAttr.cpp:3179]: It looks like this unrelated whitespace change can be reverted.

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 6]
```diff
@@ -3148,14 +3148,13 @@ static void handleWorkGroupSize(Sema &S, Decl *D, const ParsedAttr &AL) {
...
-                                     S.Context.IntTy, AL.getLoc())
-            : nullptr;
-    return E;
+    assert((AL.getKind() == ParsedAttr::AT_ReqdWorkGroupSize && AL.hasScope() &&
+            (AL.getScopeName()->isStr("sycl") ||
+             AL.getScopeName()->isStr("intel"))) &&
+           "Attribute does not exist in sycl:: or intel:: scope");
+
+    return (IntegerLiteral::Create(S.Context, llvm::APInt(32, 1),
+                                   S.Context.IntTy, AL.getLoc()));
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaDeclAttr.cpp:3157]:
```suggestion
    return IntegerLiteral::Create(S.Context, llvm::APInt(32, 1),
                                   S.Context.IntTy, AL.getLoc());
```
Extra parens don't change the order of operations, so they can be removed.

**Review by premanandrao — DISMISSED** [review]
LGTM

**Comment by schittir** [comment]
Fixing commit message now.

**Review by AaronBallman — APPROVED** [review]
LGTM! Can you also add "NFC" to the review title as this has no functional change?

**Comment by schittir** [comment]
Thank you for the review, @erichkeane  @AaronBallman - I updated the description and title. 

Thanks @premanandrao

**Comment by schittir** [comment]
Working on reproducing the test failures locally.

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 7]
```diff
@@ -3177,6 +3177,8 @@ static void handleWorkGroupSize(Sema &S, Decl *D, const ParsedAttr &AL) {
 
   ASTContext &Ctx = S.getASTContext();
 
+  assert(YDimExpr && "YDimExpr cannot be NULL");
+  assert(ZDimExpr && "ZDimExpr cannot be NULL");
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaDeclAttr.cpp:3181]: This is a bit different than the previous suggestion:
```
Agreed; I think the Klokwork report is a false positive, but we could eliminate it (hopefully) and improve the readability of the code by using an assert within the lambda to assert that the only way to call the lambda is for an attribute that can elide the Y and Z arguments. Then we can remove the return of nullptr from the lambda and simplify that code a bit.
```
I think the `assert` should be in the `SetDefault` lambda. It should be asserting:
```
(AL.getKind() == ParsedAttr::AT_ReqdWorkGroupSize && AL.hasScope() &&
         (AL.getScopeName()->isStr("sycl") ||
          AL.getScopeName()->isStr("intel")))
```
is true rather than using a conditional for it. Then we can remove the code path by which the lambda returns `nullptr` and Klokwork should be pleased again.

Comment (schittir) [reply clang/lib/Sema/SemaDeclAttr.cpp:3181]: The removal of the nullptr and adding the assert (such as in [this](https://github.com/intel/llvm/pull/4367/commits/e9f8624b342238dd9fda95e63bd70b661c545b92) commit) has a side effect or two- 
1. it trips up the error checking for the 3 argument requirement for `__attribute__((reqd_work_group_size(3)))` and `[[cl::reqd_work_group_size(3)]]` ; once the suggested assert is in `SetDefault `lambda, there is no opportunity for this check to happen at line 3174 `if (!AL.checkExactlyNumArgs(S, 3))`
2. it asserts everytime it sees a non sycl:: or intel:: usage

Comment (AaronBallman) [reply clang/lib/Sema/SemaDeclAttr.cpp:3181]: Hmmm.

I don't see how (1) is correct. Calling `SetDefaultValue()` has no impact on `AL`, so the same checking should happen regardless.

(2) Why would it ever see those usages in the first place? I would expect `isArgExpr()` to return true and you'd never get into `SetDefaultValue()` in that case. Put another way -- if we could have gotten there, then your new asserts will also fail.

(3) While this appeases Klokwork, it doesn't seem like the most clear approach to me. I think making it so that `SetDefaultValue()` cannot return a null pointer is a better approach -- then there's no way to accidentally get a null pointer when assertions are disabled.

I feel like I must be missing something. :-)

Comment (schittir) [reply clang/lib/Sema/SemaDeclAttr.cpp]: The OpenCL and cl:: spellings fall outside the scope of sycl:: or intel:: that the assert in SetDefaultValue, hence failing the assert. 

`(AL.getKind() == ParsedAttr::AT_ReqdWorkGroupSize && AL.hasScope() && (AL.getScopeName()->isStr("sycl") || AL.getScopeName()->isStr("intel")))`

In the latest [commit](https://github.com/intel/llvm/pull/4367/commits/7f49a73eb6bbac3d0cdeb50a7462a9ef04592ba0), I placed the error checking for requiring three arguments for OpenCL and cl:: spellings ahead of this assert i.e., before SetDefaultValue lambda. That solves all the aforementioned problems. Please let me know what you think.

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 8]
```diff
@@ -3175,6 +3152,26 @@ static void handleWorkGroupSize(Sema &S, Decl *D, const ParsedAttr &AL) {
       return;
   }
 
+  Expr *XDimExpr = AL.getArgAsExpr(0);
+
+  // If no attribute argument is specified, set the second and third argument
+  // to the default value 1, but only if the sycl:: or intel::
+  // reqd_work_group_size spelling was used.
```
Comment (smanna12) [inline clang/lib/Sema/SemaDeclAttr.cpp:3159]: @schittir,
Intel:: spelling of reqd_work_group_size attribute has been removed here: https://github.com/intel/llvm/pull/4433.

You do not need to check the intel:: spelling anymore. Please rebase your PR.

Comment (schittir) [reply clang/lib/Sema/SemaDeclAttr.cpp:3159]: Thank  you.

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 9]
```diff
@@ -3175,6 +3152,25 @@ static void handleWorkGroupSize(Sema &S, Decl *D, const ParsedAttr &AL) {
       return;
   }
 
+  Expr *XDimExpr = AL.getArgAsExpr(0);
+
+  // If no attribute argument is specified, set the second and third argument
+  // to the default value 1, but only if the sycl::reqd_work_group_size
+  // spelling was used.
+  //
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaDeclAttr.cpp:3160]:
```suggestion
```

**Review by AaronBallman — APPROVED** [review]
LGTM aside from a microscopic commenting nit, thanks!

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 7 above)
Comment (AaronBallman) [reply clang/lib/Sema/SemaDeclAttr.cpp]: Your new changes look great!

**Comment by schittir** [comment]
Removed extra `//` and updated description. @AaronBallman Could you please take a look and approve again? Thank you.

**Review by smanna12 — APPROVED** [review]
Thanks for the patch.

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Comment by schittir** [comment]
> LGTM!

Thank you for all the reviews!

**Comment by dm-vodopyanov** [comment]
@schittir, can you add a test for these changes?

**Comment by schittir** [comment]
> @schittir, can you add a test for these changes?

I don't think there is a real case where the assert is false. This is a non-functional change made to appease Klocwork.

**Review by MrSidims — COMMENTED** [review]
Post-factum LGTM
